File: convertedToIOS5/p769p787mandelbrotOperation/p769p787mandelbrotOperation/MyMandelbrotOperation.m


#import "MyMandelbrotOperation.h"


@implementation MyMandelbrotOperation  {
    CGSize size;
    CGPoint center;
    CGFloat zoom;
    CGContextRef bitmapContext;
}

- (id) initWithSize: (CGSize) sz center: (CGPoint) c zoom: (CGFloat) z {
    self = [super init];
    if (self) {
        self->size = sz;
        self->center = c;
        self->zoom = z;
    }
    return self;
}

- (void) dealloc {
    if (self->bitmapContext)
        CGContextRelease(self->bitmapContext);
}

- (CGContextRef) bitmapContext {
    return self->bitmapContext;
}

// create (and memory manage) instance variable
- (void) makeBitmapContext:(CGSize)sizze {
    if (self->bitmapContext)
        CGContextRelease(self->bitmapContext);
	int bitmapBytesPerRow = (sizze.width * 4);
	bitmapBytesPerRow += (16 - bitmapBytesPerRow%16)%16;
	CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    CGContextRef context = NULL;
	context = CGBitmapContextCreate(NULL, sizze.width, sizze.height, 8, bitmapBytesPerRow, colorSpace, kCGImageAlphaPremultipliedLast);
	CGColorSpaceRelease(colorSpace);
    self->bitmapContext = context;
}

#define MANDELBROT_STEPS	200

// draw pixels of self->bitmapContext

BOOL isInMandelbrotSet(float re, float im)
{
	float x = 0;	float nx;
	float y = 0;	float ny;
	bool fl = TRUE;
	for(int i = 0; i < MANDELBROT_STEPS; i++)
	{
		nx = x*x - y*y + re;
		ny = 2*x*y + im;
		if((nx*nx + ny*ny) > 4)
		{
			fl = FALSE;
			break;
		}
		x = nx;
		y = ny;
	}
	return fl;
}

- (void)drawAtCenter:(CGPoint)centerr zoom:(CGFloat)zoomm
{
	CGContextSetAllowsAntialiasing(bitmapContext, FALSE);
    CGContextSetRGBFillColor(bitmapContext, 0.0f, 0.0f, 0.0f, 1.0f);
	
	CGFloat re;
	CGFloat im;
	
    int maxi = self->size.width;
    int maxj = self->size.height;
	for (int i = 0; i < maxi; i++)
	{
		for (int j = 0; j < maxj; j++)
		{
			re = (((CGFloat)i - 1.33f * centerr.x)/160);	
			im = (((CGFloat)j - 1.00f * centerr.y)/160);	
			
			re /= zoomm;
			im /= zoomm;
			
			if (isInMandelbrotSet(re, im))
			{
				CGContextFillRect (bitmapContext, CGRectMake(i, j, 1.0f, 1.0f));
			}
		}
	}
}

- (void) main {
    if ([self isCancelled])
        return;
    [self makeBitmapContext: self->size];
    [self drawAtCenter: self->center zoom: self->zoom];
    if (![self isCancelled])
        [[NSNotificationCenter defaultCenter] 
         postNotificationName:@"MyMandelbrotOperationFinished" object:self];
}


@end

File: iOS12bookExamples/bk2ch25p865mandelbrotOperation/ch38p1106mandelbrotNoThreading/MyMandelbrotOperation.swift


import UIKit

class MyMandelbrotOperation : Operation {
    private let center : CGPoint
    private let bounds : CGRect
    private let zoom : CGFloat
    private(set) var bitmapContext : CGContext! = nil
    
    static let mandelOpFinished = Notification.Name("mandelOpFinished")

    
    init(center c:CGPoint, bounds b:CGRect, zoom z:CGFloat) {
        self.center = c
        self.bounds = b
        self.zoom = z
        super.init()
    }
    
    let MANDELBROT_STEPS = 1000

    // create instance variable
    func makeBitmapContext(size:CGSize) {
        var bitmapBytesPerRow = Int(size.width * 4)
        bitmapBytesPerRow += (16 - (bitmapBytesPerRow % 16)) % 16
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        let prem = CGImageAlphaInfo.premultipliedLast.rawValue
        let context = CGContext(data: nil, width: Int(size.width), height: Int(size.height), bitsPerComponent: 8, bytesPerRow: bitmapBytesPerRow, space: colorSpace, bitmapInfo: prem)
        self.bitmapContext = context
    }
    
    func draw(center:CGPoint, bounds:CGRect, zoom:CGFloat) {
        func isInMandelbrotSet(_ re:Float, _ im:Float) -> Bool {
            var fl = true
            var (x, y, nx, ny) : (Float,Float,Float,Float) = (0,0,0,0)
            for _ in 0 ..< MANDELBROT_STEPS {
                nx = x*x - y*y + re
                ny = 2*x*y + im
                if nx*nx + ny*ny > 4 {
                    fl = false
                    break
                }
                x = nx
                y = ny
            }
            return fl
        }
        self.bitmapContext.setAllowsAntialiasing(false)
        self.bitmapContext.setFillColor(red: 0, green: 0, blue: 0, alpha: 1)
        var re : CGFloat
        var im : CGFloat
        let maxi = Int(bounds.size.width)
        let maxj = Int(bounds.size.height)
        for i in 0 ..< maxi {
            for j in 0 ..< maxj {
                re = (CGFloat(i) - 1.33 * center.x) / 160
                im = (CGFloat(j) - 1.0 * center.y) / 160
                
                re /= zoom
                im /= zoom
                
                if (isInMandelbrotSet(Float(re), Float(im))) {
                    self.bitmapContext.fill (CGRect(CGFloat(i), CGFloat(j), 1.0, 1.0))
                }
            }
        }
    }
    
    override func main() {
        guard !self.isCancelled else {return}
        self.makeBitmapContext(size:self.bounds.size)
        self.draw(center: self.center, bounds: self.bounds, zoom: self.zoom)
        if !self.isCancelled {
            NotificationCenter.default.post(name: MyMandelbrotOperation.mandelOpFinished, object: self)
        }
    }

}
